### PR TITLE
Generalization of distributed test cases for non CUDA devices

### PR DESCRIPTION
**Pull request by AnantGulati**
## Motivation
This pr is an extension of #131758. As described in #131758, these changes are looking to make distributed UTs more accessible to users of all device types. 

The UTs currently are specific to CUDA devices as explicit CUDA API calls are used.

To allow for easier changes  to be made for non CUDA devices, we introduce a new class DistributedTestBase derived from MultiProcessTestCase  which helps abstracts out the process group creation /deletion and other functionality for a given device.

The tests can be instantiated per device using existing utilities such as instantiate_device_type_tests , which will pass the device as argument and accordingly create the PG with the right backend

The new generalized content can be added by deriving from this base class.

One such change has also been raised in this pr specifically the test_functional_api.py file. It has allowed for clean and easy adaptations for non CUDA devices as demonstrated.

CC: @kwen2501 @wconstab @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

File: torch/testing/_internal/common_distributed.py [hunk 1]
```diff
@@ -79,6 +81,7 @@ class TestSkip(NamedTuple):
         86, "Test skipped at subprocess level, look at subprocess log for skip reason"
     ),
     "importerror": TestSkip(88, "Test skipped due to missing import"),
+    "no_hpu": TestSkip(88, "HPU is not available."),
```
Comment (kwen2501) [inline torch/testing/_internal/common_distributed.py:84]: Any reason for using the same skip code?

File: torch/testing/_internal/common_distributed.py [hunk 2]
```diff
@@ -97,7 +100,10 @@ class DistTestCases:
     backend_feature["cuda"] = {"nccl", "gloo", "ucc"}
     backend_feature["ddp"] = {"nccl", "gloo", "ucc"}
     backend_feature["subgroup"] = {"nccl", "gloo", "ucc"}
+    backend_feature["hpu"] = {"hccl"}
     backend_feature["plugin"] = set()
+    if TEST_HPU:
+        backend_feature["hpu"] = {"hccl"}
```
Comment (kwen2501) [inline torch/testing/_internal/common_distributed.py:106]: Duplicated setting.

Comment (kwen2501) [reply torch/testing/_internal/common_distributed.py:106]: Also cc @wesbland @pavanbalaji on the name.

File: torch/testing/_internal/common_distributed.py [hunk 3]
```diff
@@ -48,6 +49,7 @@
 )
 import operator
 
+import habana_frameworks.torch as ht
```
Comment (kwen2501) [inline torch/testing/_internal/common_distributed.py:52]: I am not so sure about this direct import. Is there a better way to do it?

File: torch/testing/_internal/common_distributed.py [hunk 4]
```diff
@@ -106,11 +112,13 @@ def skip_if_no_gpu(func):
 
     @wraps(func)
     def wrapper(*args, **kwargs):
-        if not torch.cuda.is_available():
-            sys.exit(TEST_SKIPS["no_cuda"].exit_code)
+        if not ht.hpu.is_available():
+            sys.exit(TEST_SKIPS["no_hpu"].exit_code)
```
Comment (kwen2501) [inline torch/testing/_internal/common_distributed.py:116]: Why removing the cuda lines?

File: torch/testing/_internal/common_distributed.py [hunk 5]
```diff
@@ -172,15 +180,13 @@ def wrapper(*args, **kwargs):
...
-def at_least_x_gpu(x):
-    return torch.cuda.is_available() and torch.cuda.device_count() >= x
-
-
 def skip_if_lt_x_gpu(x):
     def decorator(func):
         @wraps(func)
         def wrapper(*args, **kwargs):
-            if torch.cuda.is_available() and torch.cuda.device_count() >= x:
+            if ht.hpu.is_available() and ht.hpu.device_count() >= x:
```
Comment (kwen2501) [inline torch/testing/_internal/common_distributed.py:187]: Same.

File: torch/testing/_internal/common_distributed.py [hunk 6]
```diff
@@ -1322,10 +1365,10 @@ def tearDown(self):
 
     @property
     def world_size(self) -> int:
-        return torch.cuda.device_count()
+        return ht.hpu.device_count()
```
Comment (kwen2501) [inline torch/testing/_internal/common_distributed.py:1368]: Same

**Review by kwen2501 — COMMENTED** [review]
Thanks for the PR. 
I didn't look closely, but it seems there are a lot of **replacements** from cuda to hpu. I wonder if this PR is a demo and not for landing?

**Comment by AnantGulati** [comment]
Accidental push with multiple commits. Closing this PR.
Will add changes again in a clean new PR

**Pull request closed without merging**